src/openms/source/ANALYSIS/ID/AScore.cpp

```diff
@@ -60,191 +59,106 @@ namespace OpenMS
-
-    if (real_spectrum.empty() || number_of_phosphorylation_events == 1 || number_of_STY == 0)
+    
+    if (real_spectrum.empty())
     {
-      return hit;
+      return phospho;
     }
 
-    TheoreticalSpectrumGenerator spectrum_generator;
+    vector<vector<Size> > permutations(computePermutations_(sites, (Int)number_of_phosphorylation_events));
     vector<RichPeakSpectrum> th_spectra;
-
-    // produce theoretical spectra with all combinations with the number of phosphorylation events
-    if (number_of_STY < number_of_phosphorylation_events)
+    if (permutations.empty())
     {
-      number_of_phosphorylation_events = number_of_STY;
+      th_spectra = createTheoreticalSpectra_(seq_without_phospho);
     }
-
-    // determine all phospho sites
-    vector<Size> sites(getSites(without_phospho));
-
-    vector<vector<Size> > permutations(computePermutations(sites, (Int)number_of_phosphorylation_events));
-
-    th_spectra.resize(permutations.size());
-    for (Size i = 0; i < permutations.size(); ++i)
+    else
     {
-      AASequence temp(without_phospho);
-      Size permu = 0;
-      for (Size as = 0; as < temp.size(); ++as)
-      {
-        if (as == permutations[i][permu])
-        {
-          temp.setModification(as, "Phospho");
-          ++permu;
-        }
-        if (permu == permutations[i].size())
-          break;
-      }
-
-      // we mono-charge spectra
-      spectrum_generator.addPeaks(th_spectra[i], temp, Residue::BIon, 1);
-      spectrum_generator.addPeaks(th_spectra[i], temp, Residue::YIon, 1);
-      th_spectra[i].setName(temp.toString());
+      th_spectra = createTheoreticalSpectra_(permutations, seq_without_phospho);
     }
-
+        
+    // prepare real spectrum windows
     if (!real_spectrum.isSorted())
     {
       real_spectrum.sortByPosition();
     }
-
-    vector<vector<double> > peptide_site_scores(th_spectra.size());
-    vector<RichPeakSpectrum> windows_top10;
-
-    // prepare peak depth for all windows in the actual spectrum
+    vector<RichPeakSpectrum> windows_top10(peakPickingPerWindowsInSpectrum_(real_spectrum));
+    
+    // calculate peptide score for each possible phospho site permutation
+    vector<vector<double> > peptide_site_scores(calculatePermutationPeptideScores_(th_spectra, windows_top10, fragment_mass_tolerance, fragment_mass_unit_ppm));
+    
+    // rank peptide permutations ascending
+    multimap<double, Size> ranking(rankWeightedPermutationPeptideScores_(peptide_site_scores));
+    
+    phospho.setScore(ranking.rbegin()->first); // initialize score with highest peptide score (aka highest weighted score)
+    phospho.setSequence(AASequence::fromString(th_spectra[ranking.rbegin()->second].getName()));
+    phospho.setMetaValue("Search_engine_sequence", hit.getSequence().toString());
+    
+    if (number_of_phosphorylation_events == 0 || number_of_STY == 0 || number_of_STY == number_of_phosphorylation_events)
```

Comment:
don't think it is possible to move it up in order to calculate the score also for peptides without an phospho event, phospho site or where the number of phospho events equal phospho sites